includes/Core/Util/Data_Request.php

```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * Data_Request
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2019 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+/**
+ * Class Data_Request
+ *
+ * @package Google\Site_Kit\Core\Util
+ */
+class Data_Request implements \ArrayAccess {
+
+	/**
+	 * Request method.
+	 *
+	 * @var string
+	 */
+	protected $method;
+
+	/**
+	 * Request type.
+	 *
+	 * @var string
+	 */
+	protected $type;
+
+	/**
+	 * Request identifier.
+	 *
+	 * @var string
+	 */
+	protected $identifier;
+
+	/**
+	 * Request datapoint.
+	 *
+	 * @var string
+	 */
+	protected $datapoint;
+
+	/**
+	 * Request data parameters.
+	 *
+	 * @var array
+	 */
+	protected $data;
+
+	/**
+	 * Request key.
+	 *
+	 * @var string
+	 */
+	protected $key;
+
+	/**
+	 * Data_Request constructor.
+	 *
+	 * @param string     $method Request method.
+	 * @param string     $type Request type.
+	 * @param string     $identifier Request identifier.
+	 * @param string     $datapoint Request datapoint.
+	 * @param array|self $data Request data parameters.
+	 * @param string     $key Request cache key.
+	 */
+	public function __construct(
+		$method = null,
+		$type = null,
+		$identifier = null,
+		$datapoint = null,
+		$data = array(),
+		$key = null
+	) {
+		$this->method     = strtoupper( $method );
+		$this->type       = $type;
+		$this->identifier = $identifier;
+		$this->datapoint  = $datapoint;
+		$this->data       = $data instanceof self ? $data->get_data() : (array) $data;
+		$this->key        = $key;
+	}
+
+	/**
+	 * Gets the request method.
+	 *
+	 * @return string
+	 */
+	public function get_method() {
```

Comment:
Should we possibly implement a magic `__get()` method to get these properties instead of this method and the similar ones below? While your current implementation is per the IB, seeing the above code changes, I think a magic getter makes more sense with interoperability of the datasets passed in a batch request. We'd be able to revert quite a few code changes specifically in the `get_batch_data()` method.

Reply:
That's an interesting idea. Would you prefer to have the properties only exposed via the magic getter or keep the getter methods as well?

Reply:
Let's do only the magic getters then (and probably `__isset()` for that purpose too)